### PR TITLE
Check user provided Twitter credentials are valid

### PR DESCRIPTION
**Pull request by fjvallarino**
## Description
Calls `account/verify_credentials` to check the credentials provided by the user are still valid.

This is a draft PR, since it's missing unit tests. I wanted to get your feedback before moving forward, mainly validating you agree with what's done and to be sure I'm not missing some expected functionality.

## Motivation and Context
It's part of our epic

## How Has This Been Tested?
It can be tested with:

- Non existent/invalid: http://services-stage.risevision.com/twitter/verify-credentials/?companyId=bad-token
- Expired: http://services-stage.risevision.com/twitter/verify-credentials/?companyId=87977ab8-38b6-47fb-ad5e-256b8cc4b46d
- Successful: http://services-stage.risevision.com/twitter/verify-credentials/?companyId=121ae986-1b89-40c8-b10b-a64c60fb9e03

You can also associate your Twitter account on any staging environment (it needs a new company, since the main Rise Vision one already has an account associated. After that, to check the token has correctly expired, you can go to `https://twitter.com/settings/applications` and revoke access to Rise Vision.

## Release Plan:
- As the Submitter, upon requesting review of this pull request, I confirm that the [Release Checklist](https://help.risevision.com/hc/en-us/articles/360031119991) has been completed. 
- As the Reviewer, upon approving the changes in this PR, I confirm I have reviewed and I agree that the [Release Checklist](https://help.risevision.com/hc/en-us/articles/360031119991) has been completed

#### Release Checklist Items Skipped?
No

@santiagonoguez @stulees

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -28,6 +28,11 @@ jobs:
           command: cp ${HOME}/gcloud-service-key.json src/gcloud-service-key.json
       - run: gcloud auth activate-service-account --key-file ${HOME}/gcloud-service-key.json
       - run: gcloud config set project $PROJECT_ID
+      - run: mkdir -p ~/.ssh
+      - run: ssh-keyscan -H github.com >> ~/.ssh/known_hosts
+      - run: git clone git@github.com:Rise-Vision/private-keys.git
+      - run: mv private-keys ..
+      - run: cp ../private-keys/twitter/twitter-app-credentials.js src/twitter
```
Comment (santiagonoguez) [inline .circleci/config.yml]: @fjvallarino it's ok, but may be nicer to group some of these commands as inline commands with pipe, and provide a name: that documents what they do ( for example create a run entry with name "get private keys repository" ) and another ( "save twitter app credentials into project" ), something like that. So the code is not too hard to read by other developers.

File: src/credentials/index.js [hunk 2]
```diff
@@ -22,14 +24,12 @@ const handleError = (res, error, errorMessage) => {
...
-      // temporarily respond with success
+    .then(twitter.verifyCredentials)
+    .then(() => {
       res.json({success: true});
-
-      // TODO: validate credentials
     })
     .catch(error=>{
-      if (error.message.includes("No credentials for")) {
+      if (error.message && (error.message.includes("No credentials for") || error.message === "Invalid or expired token.")) {
```
Comment (santiagonoguez) [inline src/credentials/index.js]: @fjvallarino just a minor suggestion, you might want to turn this long verification into a function.

File: src/twitter/index.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
+const Twitter = require('twitter');
+const TwitterCredentials = require('./twitter-app-credentials');
```
Comment (stulees) [inline src/twitter/index.js]: @fjvallarino Later we will need to remember to add documentation to README about ensuring to target the file from private keys and add to the right directory for local testing.

**Comment by stulees** [comment]
@fjvallarino Looks good! I also like @santiagonoguez suggestion about breaking up into 2 possible PRs (nodemon, validate credentials).

File: .circleci/config.yml (hunk 1 above)
Comment (fjvallarino) [reply .circleci/config.yml]: @santiagonoguez sounds good!

File: src/credentials/index.js (hunk 2 above)
Comment (fjvallarino) [reply src/credentials/index.js]: @santiagonoguez makes sense!

File: src/twitter/index.js (hunk 3 above)
Comment (fjvallarino) [reply src/twitter/index.js]: @stulees do you mean documenting that we're grabbing the keys from `private-keys` for deployment?

Comment (stulees) [reply src/twitter/index.js]: @fjvallarino No I'm referring more to running the service locally. Won't there be an error if this file is missing?

Comment (fjvallarino) [reply src/twitter/index.js]: @stulees good point. Do you think it's worth having a dummy file with invalid credentials?

Comment (stulees) [reply src/twitter/index.js]: @fjvallarino Yeah, the file could even be empty (`module.exports = {}`) with commented out instructions.

**Comment by fjvallarino** [comment]
@santiagonoguez @stulees this is ready for review again. Most of the suggestions/change requests were applied, saving the replacement of `otp-redis-master` for `otp-redis-slave`, since it was failing. Maybe we can try a separate PR for that.

File: src/twitter/index.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  });
+
+  return client;
+};
+
+const verifyCredentials = credentials => {
+  return invokeEndpoint(credentials, 'account/verify_credentials');
+};
+
+const invokeEndpoint = (clientCredentials, endpoint, args) => {
```
Comment (santiagonoguez) [inline src/twitter/index.js]: @fjvallarino this function could be reused by get-tweets, right ?

Comment (fjvallarino) [reply src/twitter/index.js]: @santiagonoguez yes, it could! I will make a small refactoring in order for it to return the response instead of just true.

Comment (fjvallarino) [reply src/twitter/index.js]: @santiagonoguez done!

**Comment by stulees** [comment]
@fjvallarino @santiagonoguez FYI the changes to support CORs are in the last commit. I've tested my Apps changes against our staging service and it now worked.

**Comment by fjvallarino** [comment]
@stulees nice! I'll merge now

**Pull request merged**